src/riak_cs_simple_quota.erl

```diff
@@ -0,0 +1,234 @@
+    {ok, RD, Ctx}.
+
+-spec maybe_usage(binary(), #context{}) -> non_neg_integer().
+maybe_usage(_, _Ctx = #context{riak_client=undefined}) ->
+    %% can't happen here
+    error(no_riak_client);
+maybe_usage(User0, _Ctx = #context{riak_client=RiakClient}) ->
+    User = binary_to_list(User0),
+    Usage = case get_latest_usage(RiakClient, User) of
+                {error, notfound} ->
+                    lager:warning("No storage stats data was found. Starting as no usage."),
+                    0;
+                {ok, Usages} ->
+                    sum_all_buckets(lists:last(Usages))
+            end,
+    %% Here's a race condition where if so many concurrent access
+    %% come, each access can yield a new fresh state and then
+    %% receives not access limitation.
+    try
+        ets:insert_new(?MODULE, {User0, Usage}),
+        Usage
+    catch _:_ ->
+            Usage
+    end.
+
+update(User,
+       #wm_log_data{method = Method,
+                    headers = Headers} = _LogData)
+  when Method =:= 'PUT' orelse Method =:= 'DELETE' ->
+
+    RequestLength = mochiweb_headers:get_value("content-length", Headers),
+    Bytes = case catch list_to_integer(RequestLength) of
+                Len when is_integer(Len) ->
+                    case Method of
+                        'PUT' -> Len;
+                        'DELETE' -> - Len
+                    end;
+                _ -> 0
+            end,
+    try
+        ets:update_counter(?MODULE, User, Bytes)
+    catch Type:Error ->
+            %% TODO: show out stacktrace heah
+            lager:warning("something wrong? ~p", [Error]),
+            {error, {Type, Error}}
+    end;
+update(_, _) ->
+    ok.
+
+error_response({disk_quota, Current, Limit}, RD, Ctx) ->
+    StatusCode = 403,
+    XmlDoc = {'Error',
+              [
+               {'Code', [StatusCode]},
+               {'Message', ["You have Exceeded your quota. Please delete your data."]},
+               {'CurrentValueOfQuota', [Current]},
+               {'AllowedLimitOfQuota', [Limit]}
+              ]},
+    Body = riak_cs_xml:to_xml([XmlDoc]),
+    ReqData = wrq:set_resp_header("Content-Type", ?XML_TYPE, RD),
+    UpdReqData = wrq:set_resp_body(Body, ReqData),
+    {StatusCode, UpdReqData, Ctx}.
+
+-spec get_latest_usage(pid(), string()) -> {ok, list()} | {error, notfound}.
+get_latest_usage(Pid, User) ->
+    Now = calendar:now_to_datetime(os:timestamp()),
+    NowASec = calendar:datetime_to_gregorian_seconds(Now),
+    get_latest_usage(Pid, User, NowASec, 0, 10).
+
+-spec get_latest_usage(pid(), string(),
+                       non_neg_integer(), non_neg_integer(), non_neg_integer()) ->
+                              {ok, list()} | {error, notfound}.
+get_latest_usage(_Pid, _User, _, N, N) -> {error, notfound};
+get_latest_usage(Pid, User, EndSec, N, Max) ->
+    End = calendar:gregorian_seconds_to_datetime(EndSec),
+    ADayAgo = calendar:gregorian_seconds_to_datetime(EndSec - 86400),
```

Comment:
If hard coded `86400` is meant for the default value of `storage_archive_period`, using application environment value would be better.